cosima_cookbook/database.py

```diff
@@ -579,37 +545,98 @@ def build_index(
-    Symbolically linked files and/or directories will be indexed if followsymlinks is True.
+    Requires a session for the database that's been created with the
+    create_session() function. If client is not None, use a distributed
+    client for processing files in parallel. If prune is set to 'flag' files
+    that are already in the database but are missing from the filesystem will
+    be flagged as missing, or if set to 'delete' they will be removed from the
+    database. If force is False only files that are not in the database will
+    be indexed, othewise if True all files will be indexed and their database
+    entries updated. Symbolically linked files and/or directories will be
+    indexed if followsymlinks is True.
 
     Returns the number of new files that were indexed.
     """
 
     if not isinstance(directories, list):
         directories = [directories]
 
-    indexed = 0
-    for directory in directories:
-        indexed += index_experiment(
-            directory, session, client, update, prune, delete, followsymlinks
+    prune = prune.lower()
+    if not prune in {"flag", "delete"}:
+        print(
+            "build_index :: ERROR :: Value for option prune is incorrect,\n "
+            "must be flag or delete: {}\n"
+            "Resetting to 'flag'".format(prune)
         )
+        prune = "flag"
+
+    indexed = 0
+    for directory in [Path(d) for d in directories]:
+        # find all netCDF files in the experiment directory
+        files = find_files(directory, followsymlinks=followsymlinks)
+        expt = find_experiment(session, directory)
+
+        if expt is None:
+            expt = NCExperiment(
+                experiment=str(directory.name), root_dir=str(directory.resolve())
+            )
+
+        print("Indexing experiment: {}".format(directory.name))
+
+        existing_files = set([f.ncfile for f in expt.ncfiles])
+
+        # Prune files that exist in the database but are not present
+        # on the filesystem
+        missing_files = existing_files - files
+        if len(missing_files) > 0:
+            # Make a list of NCFile objects to pass to prune_files
+            missing_files_objs = [f for f in expt.ncfiles if f.ncfile in missing_files]
+            prune_files(expt, session, missing_files_objs, delete=(prune == "delete"))
+
+        if force:
+            # Delete entries that we're reindexing
+            for f in files & existing_files:
+                existing_ncfile = (
+                    session.query(NCFile)
+                    .join(NCExperiment)
+                    .filter(NCExperiment.experiment == str(directory.name))
+                    .filter(NCExperiment.root_dir == str(directory.resolve()))
+                    .filter(NCFile.ncfile == f)
+                    .one_or_none()
+                )
+                if existing_ncfile is not None:
+                    print("Re-indexing {}".format(f))
+                    session.delete(existing_ncfile)
```

Comment:
I was trying to reuse the file list because I didn't know how to do this nice and succinctly as you've done.

I think I'm getting closer and closer to what you originally said we should do.

I'm not too bothered about adding an event.